File: src/controllers/todoController.js
const TodoModel = require('../models/todoModel');

class TodoController {
  static sendError(res, statusCode, message, details) {
    const payload = { error: message };

    if (details !== undefined) {
      payload.details = details;
    }

    return res.status(statusCode).json(payload);
  }

  // GET /api/todos
  static getTodos(req, res) {
    const { q, completed } = req.query;
    const todos = TodoModel.getAll({ q, completed });
    res.json(todos);
  }

  // GET /api/todos/stats
  static getTodoStats(req, res) {
    res.json(TodoModel.getStats());
  }

  // GET /api/todos/:id
  static getTodoById(req, res) {
    const todo = TodoModel.getById(req.params.id);
    if (!todo) {
      return this.sendError(res, 404, 'Todo not found');
    }
    res.json(todo);
  }

  // POST /api/todos
  static createTodo(req, res) {
    const { title, completed } = req.body;

    const newTodo = TodoModel.create({ title, completed });
    if (!newTodo) {
      return this.sendError(res, 400, 'Title is required', {
        field: 'title',
        received: title
      });
    }

    res.status(201).json(newTodo);
  }

  // PUT /api/todos/:id
  static updateTodo(req, res) {
    const { title, completed } = req.body;
    const updatedTodo = TodoModel.update(req.params.id, { title, completed });

    if (!updatedTodo) {
      if (title !== undefined && TodoModel.normalizeTitle(title) === null) {
        return this.sendError(res, 400, 'Title cannot be empty', {
          field: 'title',
          received: title
        });
      }

      return this.sendError(res, 404, 'Todo not found');
    }

    res.json(updatedTodo);
  }

  // POST /api/todos/:id/toggle
  static toggleTodo(req, res) {
    const toggledTodo = TodoModel.toggle(req.params.id);

    if (!toggledTodo) {
      return this.sendError(res, 404, 'Todo not found');
    }

    res.json(toggledTodo);
  }

  // DELETE /api/todos/:id
  static deleteTodo(req, res) {
    const deletedTodo = TodoModel.delete(req.params.id);
    if (!deletedTodo) {
      return this.sendError(res, 404, 'Todo not found');
    }

    res.json({ message: 'Todo deleted successfully', deletedTodo });
  }

  // DELETE /api/todos/completed
  static clearCompletedTodos(req, res) {
    const removedTodos = TodoModel.clearCompleted();

    res.json({
      message: 'Completed todos cleared successfully',
      removedCount: removedTodos.length,
      removedTodos
    });
  }
}

module.exports = TodoController;

File: src/controllers/todo_controller.py
from fastapi import HTTPException, status
from typing import Optional
from pydantic import BaseModel
from src.models.todo_model import TodoModel

class CreateTodoRequest(BaseModel):
    title: str
    completed: Optional[bool] = False

class UpdateTodoRequest(BaseModel):
    title: Optional[str] = None
    completed: Optional[bool] = None

class TodoController:
    @staticmethod
    def get_todos(q: Optional[str] = None, completed: Optional[str] = None):
        return TodoModel.get_all(q, completed)

    @staticmethod
    def get_todo_by_id(id: int):
        todo = TodoModel.get_by_id(id)
        if not todo:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Todo not found")
        return todo

    @staticmethod
    def create_todo(body: CreateTodoRequest):
        if not body.title:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Title is required")
        return TodoModel.create(body.title, body.completed)

    @staticmethod
    def update_todo(id: int, body: UpdateTodoRequest):
        updated = TodoModel.update(id, body.title, body.completed)
        if not updated:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Todo not found")
        return updated

    @staticmethod
    def delete_todo(id: int):
        deleted = TodoModel.delete(id)
        if not deleted:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Todo not found")
        return {"message": "Todo deleted successfully", "deletedTodo": deleted}
